### ISSUE TITLE
Generalize SetItemsProcessed and SetBytesProcessed

### ISSUE THREAD

**Issue by baryluk** [google/benchmark#324]
Simple feature frequest.

It would be nice to say something like this:

    SetProcessed("requests", 123, "QPS");
    SetProcessed("keys", 12300, "KPS");
    SetProcessed("bytes", 90000, "B/s");
    SetProcessed("errors", 1, "EPS");

etc.

Benchmark framework would sort them by key (or just use in specified order) in each benchmark, compute rates (taking into accounting wall time or cpu time as needed, possibly configurable per SetProcessed call), separate them by semicolons, and if needed format them properly for human consumption.

Proposal:

    SetProcessed(string key, int64 processed, string unit, Options options = Options());

Options could have things like: should average, min, max or sum, be taken from different threads; should we use cpu time or wall time clock, should k, M, G, prefixes be used, or maybe floating point with exponent, or just integer with fixed rounding, should additional statistic be compute (percentiles, median, std dev, iqr, normality test - i.e. from different benchmark repetitions, etc).

Key would be used in other formats like json, csv, to name columns.

We can also extend Benchmark, to allow extending columns explicitly, i.e.

    ->AddProcessedKey("requests", 10)
    ->AddProcessedKey("keys", 10)
    ->AddProcessedKey("bytes")

This could be used by the console to make nice table, including a table header row. 10 in an example above would be optional column width.

Not sure, what would be approach, if different benchmarks declare different key names. (Reemit new header, if it is different than a previous one?)

**Comment by baryluk** [comment]
BTW. SetLabel would remain unchanged, and still be printed after all key rates.

**Comment by dominichamon** [comment]
this is #262 i think?

**Comment by baryluk** [comment]
Yes indeed. I was not aware of that PR, or issue #240 

There are few differences in my proposal, i.e. ability to report a rates (not just counters, which doesn't play well when using with automatic iteration count computation or time based iterations), and ability to set options and suffixes for example

**Comment by dominichamon** [comment]
perhaps it can be extended to cover your use cases (once it lands.. longest PR ever)?

**Comment by LebedevRI** [comment]
Question: is there any status here?
I would like to have a median in addition to the currently-available mean.
If i just add it, would that be upstreamable, or is the full rework like suggested here is the only way to go?

**Comment by LebedevRI** [comment]
Edit: i didn't know `Stat1` was optimized to not keep the values themselves...
I did implement median, but ATM at the cost of **adding** `std::vector<VType> values_;`
It is probably too intrusive?

**Comment by dominichamon** [comment]
I don't think it's intrusive. In fact, calculating the mean, stddev, median, etc at report time is probably the right thing to do.

**Comment by LebedevRI** [comment]
Ok, so if i refactor `Stat1` to store `std::vector<VType>`, that would be fine?

**Comment by dominichamon** [comment]
I think so. `Stat1` is only used after the benchmark has run so any concerns about `vector` performance should be minimal. I would like to see some thought put into is how large the vectors could be as the number of runs increase.

**Comment by LebedevRI** [comment]
@baryluk i think this is resolved, now that we have custom user counters?
The nicer support for custom units _may_ be nice to have, but not too impactful.

**Comment by louchenyao** [comment]
@LebedevRI Hello, based on my understanding, `ComputeStatistics` passed by a custom lambda function, which only accepts the running time of iterations. But I think we need the `state.range(0)` to compute the processed items for each iteration.

**Comment by LebedevRI** [comment]
You want to know "per one iteration, how many items are processed", correct?
So just accumulate the `ItemsTotal` in some variable that is outside of the workloop,
and after the workloop just do
```
state.counters["ItemsPerIteration"] = Counter(ItemsTotal, benchmark::Counter::kAvgIterations);
```
By definition, 
https://github.com/google/benchmark/blob/7d97a057e16597e8020d0aca110480fe82c9ca67/include/benchmark/benchmark.h#L367-L369
seems quite exactly what you are looking for.

**Comment by LebedevRI** [comment]
@louchenyao did that answer your question?

**Comment by louchenyao** [comment]
@LebedevRI Thank you! This solved my question!

**Comment by LebedevRI** [comment]
I think i'm gonna close this.
